Subtype: avoid some false alarm in `subtype_unionall`

N5N3 opened the pull request:
The current check is not correct if we set the typevar's bounds to an unwrapped `UnionAll`. Use `var_occurs_inside_skip` to skip the inside check.
close #24333, close #47654

vtjnash commented:
@nanosoldier `runtests(ALL, vs=":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47868#issuecomment-1353762667) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/958adf5_vs_847cdde/report.html).

N5N3 committed:
Subtype: avoid some false alarm in `subtype_unionall`

The current check is not correct if we set the typevar's bounds to an unwrapped `UnionAll`. Use `var_occurs_inside_skip` to skip the inside check.

N5N3 committed:
Revert "Subtype: avoid some false alarm in `subtype_unionall`"

This reverts commit b5f79828d0b071722341b29e07d510359810847e.

N5N3 committed:
Make sure `UnionAll` keeps wrapped when bounds setting.

Also add `env` test.

N5N3 commented:
Failure in `Unitful` shows that [b5f7982](https://github.com/JuliaLang/julia/pull/47868/commits/b5f79828d0b071722341b29e07d510359810847e) fails to pick the right env.
So perhaps we'd better avoid the `UnionAll` unwrapping instead.

N5N3 commented:
@nanosoldier runtests(ALL, vs=":master")

N5N3 commented:
@nanosoldier `runtests(ALL, vs=":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47868#issuecomment-1355215153) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/2bac38a_vs_b6f32bc/report.html).

N5N3 commented:
The result looks good. The only concern is `MosekTools`, its test passed on my PC with a local solver, but I saw some random error if I use the network fallback solver.

Since the "Segmentation fault" was happened in libmosek64, I thought this PR might be unrelated? cc @blegat @ulfworsoe

blegat commented:
I would assume it is not related to this PR Indeed, @ulfworsoe might know better where this crash is coming from, see https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/2bac38a_vs_b6f32bc/MosekTools.primary.log

vtjnash commented:
As I look further into this, I think the concept of `var_occurs_inside` and especially `in_union` is somewhat flawed. I made a counter-example of this here: https://github.com/JuliaLang/julia/pull/31167#issuecomment-1358381818

My current theory is that we need to instead deal with this by setting the var in the env to something unsatisfyable (either with a bit, or with a falsifiable constraint such as `Any <: ub <: Union{}`) whenever computing the intersection for the lb and ub

JeffBezanson commented:
I may have a fix for #47658 that also fixes these cases.
